documentation/src/main/asciidoc/user_guide/server_protocols_hotrod.adoc

```diff
@@ -104,19 +105,67 @@ assert !cache.containsKey("car") : "Value must have been removed!";
 
 The client API maps the local API: link:{javadocroot}/org/infinispan/client/hotrod/RemoteCacheManager.html[RemoteCacheManager] corresponds to link:{javadocroot}/org/infinispan/manager/DefaultCacheManager.html[DefaultCacheManager] (both implement link:{javadocroot}/org/infinispan/manager/CacheContainer.html[CacheContainer] ). This common API facilitates an easy migration from local calls to remote calls through Hot Rod: all one needs to do is switch between link:{javadocroot}/org/infinispan/manager/DefaultCacheManager.html[DefaultCacheManager] and link:{javadocroot}/org/infinispan/client/hotrod/RemoteCacheManager.html[RemoteCacheManager] - which is further simplified by the common link:{javadocroot}/org/infinispan/manager/CacheContainer.html[CacheContainer] interface that both inherit.
 
-All keys can be retrieved from the remote cache (whether it's local, replicated, or
-distributed) by using +keySet()+ method. If the remote cache is a distributed cache,
-the server will perform a distributed stream operation to retrieve all keys from
-clustered nodes, and return all keys to the client.  Please use this method with
-care if there are large number of keys.
+==== RemoteCache(.keySet|.entrySet|.values)
+
+The collection methods `keySet`, `entrySet` and `values` are backed by the remote cache.
+That is that every method is called back into the `RemoteCache`. This is useful as
+it allows for the various keys, entries or values to be retrieved lazily, and not
+requiring them all be stored in the client memory at once if the user does not want.
+These collections adhere to the `Map` specification being that `add` and `addAll`
+are not supported but all other methods are supported.
+
+One thing to note is the `Iterator.remove` and `Set.remove` or `Collection.remove`
+methods require more than 1 round trip to the server to operate. You can check
+out the link:{javadocroot}/org/infinispan/client/hotrod/RemoteCache.html[RemoteCache]
+Javadoc to see more details about these and the other methods.
+
+*Iterator Uage*
+
+The iterator method of these collections uses `retrieveEntries` internally, which
+is described below. If you nottice `retrieveEntries` takes an argument for the batch
+size. There is no way to provide this to the iterator. As such the batch size can be
+configured via system property `infinispan.client.hotrod.batch_size` or through
+the link:{javadocroot}/org/infinispan/client/hotrod/configuraion/ConfigurationBuilder.html#batchSize-int-[ConfigurationBuilder]
```

Comment:
imo configuration should be the preferred place; why do we need the system property at all? JCache? Is there a way to add proprietary configuration when using JCache?